pkg/reconciler/managed/reconciler.go

```diff
@@ -83,12 +87,96 @@ const (
+	// resource is updated. Useful for immutable external resources.
+	sets.New[xpv1.ManagementAction](xpv1.ManagementActionObserve, xpv1.ManagementActionCreate, xpv1.ManagementActionDelete, xpv1.ManagementActionLateInitialize),
+	// No Update and no Delete, the external resource is not updated
+	// when the managed resource is updated and the external resource
+	// is not deleted when the managed resource is deleted.
+	sets.New[xpv1.ManagementAction](xpv1.ManagementActionObserve, xpv1.ManagementActionCreate, xpv1.ManagementActionLateInitialize),
+	// No Update and no LateInitialize, the external resource is not updated
+	// when the managed resource is updated and the spec.forProvider is not
+	// late initialized.
+	sets.New[xpv1.ManagementAction](xpv1.ManagementActionObserve, xpv1.ManagementActionCreate, xpv1.ManagementActionDelete),
+	// No Update, no Delete and no LateInitialize, the external resource is
+	// not updated when the managed resource is updated, the external resource
+	// is not deleted when the managed resource is deleted and the
+	// spec.forProvider is not late initialized.
+	sets.New[xpv1.ManagementAction](xpv1.ManagementActionObserve, xpv1.ManagementActionCreate),
+}
+
 // ControllerName returns the recommended name for controllers that use this
 // package to reconcile a particular kind of managed resource.
 func ControllerName(kind string) string {
 	return "managed/" + strings.ToLower(kind)
 }
 
+// ManagementActionChecker is used to perform checks on management actions
+// to determine if they are allowed, or if they are the only allowed action.
+type ManagementActionChecker interface {
+	// ActionAllowed returns true if the supplied action is allowed.
+	ActionAllowed(action xpv1.ManagementAction) bool
+	// OnlyAction returns true if the supplied action is the only allowed action.
+	OnlyAction(action xpv1.ManagementAction) bool
+}
+
+// ManagementPolicyActionChecker is used to perform checks on management actions
+// based on the management policy and if the management policy feature is enabled.
+type ManagementPolicyActionChecker struct {
+	managementPolicyEnabled bool
+	managementPolicy        xpv1.ManagementPolicy
+}
+
+// NewManagementPolicyActionChecker returns an ActionChecker based on the
+// management policy and if the management policy feature is enabled.
+func NewManagementPolicyActionChecker(managementPolicyEnabled bool, managementPolicy xpv1.ManagementPolicy) *ManagementPolicyActionChecker {
+	return &ManagementPolicyActionChecker{
+		managementPolicyEnabled: managementPolicyEnabled,
+		managementPolicy:        managementPolicy,
+	}
+}
+
+// ActionAllowed returns true if the supplied action is contained in the
+// management policy. Returns true if the management policy feature is disabled.
+func (m *ManagementPolicyActionChecker) ActionAllowed(action xpv1.ManagementAction) bool {
+	if !m.managementPolicyEnabled {
+		return true
+	}
+	return ContainsManagementAction(m.managementPolicy, action)
```

Comment:
Should we just inline these functions into their matching methods? Do they need to exist separately?

Reply:
They were used both in the reconciler and as helper functions for the providers. I removed them now as they are not used anymore in the reconciler. And I think the check will be simple enough to not need a helper function for the providers here.